lib/semian/circuit_breaker.rb

```diff
@@ -81,32 +91,39 @@ def half_open?
     def half_open
       log_state_transition(:half_open)
       @state = :half_open
-      @successes = 0
+      #@successes = 0
+      @shared_circuit_breaker_data.successes=0
     end
 
     def success_threshold_reached?
-      @successes >= @success_count_threshold
+      #@successes >= @success_count_threshold
+      @shared_circuit_breaker_data.successes >= @success_count_threshold
     end
 
     def error_threshold_reached?
-      @errors.count == @error_count_threshold
+      #@errors.count == @error_count_threshold
+      @shared_circuit_breaker_data.count == @error_count_threshold
     end
 
     def error_timeout_expired?
-      @errors.last && (@errors.last + @error_timeout < Time.now)
+      #@errors.last && (@errors.last + @error_timeout < Time.now)
+      time_f = @shared_circuit_breaker_data.last
+      time_f && (Time.at(time_f) + @error_timeout < Time.now)
     end
 
     def push_time(window, max_size, duration:, time: Time.now)
-      window.shift while window.first && window.first + duration < time
+      #window.shift while window.first && window.first + duration < time
```

Comment:
This method is racy - there needs to be a lock around `window`. You can easily have multiple process making changes to the underlying array in parallel.